### DISCUSSION TITLE
max_buffer_bytesize for async_producer

### DISCUSSION THREAD

**Discussion by alan-sapaad** [zendesk/ruby-kafka#922] [Q&A] [answered]
I am able to set `max_buffer_bytesize` for a synchronous producer and as expected I am receiving the error `Kafka::BufferOverflow: Cannot produce to topic, max buffer bytesize (960000 bytes) reached`

But I need to use an async_producer with `delivery_interval` set to `50ms (0.05)`. I've set `max_buffer_bytesize` when initializing the producer. But I am able to push as many messages as possible to the Queue and all I can see is the queue size getting bigger every second (even if I don't produce any message). I am guessing it must be the [:deliver_message, nil] packets that are being pushed into the queue every 50ms.

I am not able to receive any errors from the underlying sync producer. How do I make sure that the package size is under 1MB? I am using Heroku Kafka and they will not let any packages over 1MB. I want the `async_producer.produce` to error out if the buffer size will be more then 950kb with the current message added.

Thanks!

**Comment by dasch** [comment, marked as answer]
You need to configure the queue in addition to the buffer – but there should already be an upper bound on the queue size. The async producer runs in a different thread than the calling code, so you cannot rely on knowing the current buffer size, since the producer will be trying to flush that buffer concurrently with your calling code.

**Reply by alan-sapaad** [reply]
Got it 👍🏽 

We do rarely get "queue clogs" where messages just pile up in the async_queue. We are guessing this happens where there is a network glitch. We have seen this also when the Kafka broker is undergoing maintenance.

Our rate of massage production is very low (~25/s). To minimize these queue clogs, we have initialized 3 separate async_producers in a single dyno and when a message is produced from the application code, it picks up the producer with the smallest queue size

```
# initializing 3 producers:
$KAFKA_PRODUCER_BATCH =
    (0..2).map do |index|
      {
        id: index,
        instance: $KAFKA_CLIENT.async_producer(
          delivery_interval: 0.12,
        )
      }
    end

```

```
# select producer with smallest queue size
producer = $KAFKA_PRODUCER_BATCH.min_by { |producer| producer[:instance].instance_variable_get(:@queue).size }
```

This is working fine 99% of the time.
But very rarely, these queues get big, very rapidly. This happens even when the message production rate is minimal. Sometimes it gets cleared automatically in a couple of minutes. Sometimes it doesn't unless we restart the particular dyno. But we will lose a couple of messages in this case. Looking deep into this, we think:

1. The rapid growth of queue size is just full of `[:deliver_messages, nil]` flag appended every 120ms (our set delivery interval). Is there a way to avoid this? Like, if the last message is already `[:deliver_messages, nil]`, then do not append again? Also, would having these deliver flags hundreds of times one after another affect the delivery process, like would the network requests build on top of another when the worker thread loops through this queue full of deliver flags?
2. Is it a good idea to just kill the specific producer if this queue clog happens and spin up another programmatically? Are there existing tools, or design patterns which we could look at ? Also, is there an easy way to "copy over" all the "stuck" messages in the clogged producer to the new one?

Please share your thoughts.
Thanks!

**Reply by alan-sapaad** [reply]
I just got a reply back from the Heroku Engineer that, if I need to ensure 100% resilience, I'll have to use a synchronous producer inside a Sidekiq worker. So messages are queued in the Sidekiq queue, which is more resilient.

What do you think about this pattern?

**Reply by dasch** [reply]
I don't know why writing to a Sidekiq queue would be _more_ reliable than writing to Kafka...

I don't _think_ the `:deliver_messages` message (indicating that the time-based delivery trigger should be invoked) does anything if there are no buffered messages. The main problem is probably that the producer can't get rid of the messages by writing to the partition leader. The only thing you can really do is buffer these messages for _longer_, I think, at least if you have a requirement that the messages be written to a particular partition.

1. Can you share some logs?
2. Would it be OK for messages to be _randomly_ assigned to a partition? Because if so, just choosing another partition for the messages could be a good solution. I'm not sure if the current Producer implementation can do that, but I don't think it would be _that_ hard to implement.

**Reply by alan-sapaad** [reply]
1. We didn't have the logger enabled for the Kafka client, we just added one. We luckily haven't faced any issues after that
2. Yes, it should be ok for a message to be randomly assigned to a partition, _if a queue clog happens_.

> I don't know why writing to a Sidekiq queue would be more reliable than writing to Kafka...

Messages in the Kafka queue can be lost if a dyno restarts or if it crashes for some reason, right. But if I setup a synchronous producer or a lower level producer inside a sidekiq worker then the messages are guaranteed to be delivered because they are persisted in Redis rather than dyno-in-memory, right ?
This might not be ideal if the rate of message production is high, we will not be able to make use of message batching and all that the Kafka client does.

**Reply by dasch** [reply]
Yeah, you're just describing the difference between sync and async processing – async processing inherently has stuff lingering in memory, which can disappear for any number of reasons. If your write to redis is async, then you'll have the same problem.

**Reply by dasch** [reply]
If you don't _need_ specific partition assignment, then I would choose random assignment and an async producer, unless you are willing to fail the incoming requests if the Kafka write can't succeed. Typically you'd rather lose the Kafka message than fail the requests, though, depending on the use case.

Now, it might be that partition assignment isn't changed even after an error – but that _could_ be fixed in ruby-kafka, if you're up for writing a PR. If a broker fails and the partitioning strategy is random assignment, then repartition the messages and put them back in the buffer.

**Reply by alan-sapaad** [reply]
Sure, I will look into that.
What if I don't use a producer at all, but use `client.deliver_message()` (https://github.com/zendesk/ruby-kafka/blob/master/lib/kafka/client.rb#L155) inside a Sidekiq worker?

**Reply by dasch** [reply]
That doesn't really change anything – it's effectively just a producer with no buffering. Write throughput will be much worse.

**Reply by alan-sapaad** [reply]
But would that be more resilient? The application will be in full control of individual messages, so we can initiate retry how many ever times we want to?
Please note that the rate of message production is just around 25 per second for us

**Reply by dasch** [reply]
I don't think writing to Redis is inherently more reliable than writing to Kafka. So you're now stuck with _two_ pieces of infrastructure that may fail. Personally, I would not consider using Sidekiq for this, and I think it's terrible advice.

**Reply by alan-sapaad** [reply]
yeah, makes sense.

I am able to get the queue length by doing `producer[:instance].instance_variable_get(:@queue).size`. But like I said, if the queue is clogged, the size grows rapidly, but not with actual messages but with the flags `[:deliver_message, nil]`, because we have set the `delivery_interval` to 120ms.
1. Is there a way to avoid adding this flag to the queue, if the previous message is already a flag?
2. Or is there a way to get the queue size excluding these flags?

**Reply by dasch** [reply]
There's no _thread safe_ way to do either.

I would also say: a `delivery_interval` of 120ms is _really_ low. Have you tried adjusting that to at least 1s?

**Reply by dasch** [reply]
There might be a different design for the async producer that uses another mechanism for signaling that the delivery interval has been reached, but I'm not going to have the time to look into it any time soon.

Have you looked at using ruby-rdkafka for this? I have been meaning to migrate DeliveryBoy from ruby-kafka to ruby-rdkafka in order to have a high level API, but, yeah, I'm off doing non-Ruby, non-Kafka things, and don't have enough time to look into this stuff anymore.

**Reply by alan-sapaad** [reply]
@dasch we do have time-sensitive transactional data flowing through the system as well. We thought 120ms is like a sweet spot to have the message delivery latency as low as possible and have a hand full of messages in the batch. The faster we get the messages on the other end, the better.

No, I have not looked into ruby-rdkafka yet. I think I'm going to try it out.

Thanks again!
